### ISSUE TITLE
How to kill a zombie transaction?

### ISSUE THREAD

**Issue by chuck-h** [RWTH-i5-IDSG/steve#94]
I have an old "ghost" transaction in my database which was never correctly terminated. Is there, or can there be, a UI method for cleanly terminating it?

**Comment by chuck-h** [comment]
One approach would be to mimic a `StopTransaction.req` message from the CP, with the current timestamp and a stop reason `OTHER`. Unfortunately that leaves no specific indication of what happened. It would also be tempting to create an additional `Reason` enum for this case, but it feels like bad practice to muck with the OCPP data types.

The transaction db table could be expanded with a note field, editable from the transaction detail screen. This would provide a place to explain the forced termination; it could have other useful applications too. Transaction detail screen would probably be the right place to put a "force termination" button.

I'm not sure there is a definitively correct way to fill in the `meterStop` field in this case, especially if there were no intermediate metervalue reports.

**Comment by lincomatic** [comment]
I actually ran into this exact problem yesterday... I had about 120 zombie transactions going, because I've been using steve to debug my OCPP client, and there was no way to flush them out. I finally wrote an my CP side to loop through all the transaction ID's and send StopTransaction messages. It would be nice to have a cleaner solution.

**Comment by goekay** [comment]
@lincomatic are these transactions identical? i mean, we already have [logic](https://github.com/RWTH-i5-IDSG/steve/blob/master/src/main/java/de/rwth/idsg/steve/repository/impl/OcppServerRepositoryImpl.java#L289-L329) to prevent insterting the same StartTransaction if it arrives multiple times. why does the charging station not send the StopTransaction in the first place? the solution that i thought of was manually stopping a zombie transaction. i cannot envision an automated way of killing _120_ transactions (or preventing them). [the problems that we deal with](https://github.com/RWTH-i5-IDSG/steve/issues/97#issuecomment-445395765). _sigh_

**Comment by lincomatic** [comment]
@goekay They weren't identical. I am writing OCPP CP code, and in the process of debugging it, sometimes, a StopTransaction isn't sent out. I can see this kind of thing happening even in a live setup, for instance if the connection is flaky, or the CP gets rebooted or crashes mid transaction. There needs to be a way for the operator to get rid of zombie transactions. Two ideas I have for dealing with this:

1) have checkboxes on the transactions screen, where the operator can just select and delete them, perhaps setting a stop reason, as well.
2) from the transactions screen, the operator selects a transaction ID, and from the individual transaction screen, add a way to delete that single transaction.

**Comment by goekay** [comment]
both ideas are possible but they do not scale to _120_ zombie transactions.

the thing that i am asking is: why does the CP not queue and persist messages that it cannot send successfully (due to flaky connection, reboot, etc.) and try them again?

**Comment by lincomatic** [comment]
idea (1) scales better for multiple transactions if each screen has a checkbox on the top that if checked, checks all of the transactions currently on the screen.

Theoretically, the CP should be written to properly re-send messages when it crashes or gets rebooted, but I don't think you can be 100% sure that all CPs will be reliable and bug-free. Or say CP hardware malfunctions mid-transaction, and the CP is taken offline to be repaired or permanently removed. There are a myriad of situations where zombie transactions could creep in.

**Comment by goekay** [comment]
i have a not-final-yet approach to address this. first, i'll create the following db table for bookkeeping reasons:

```
CREATE TABLE transaction_terminated (
  transaction_pk INT NOT NULL,
  terminated_at TIMESTAMP(6) NOT NULL,
  stop_timestamp_strategy ENUM('same_as_start', 'last_meter_value', 'start_of_next_tx'),
  stop_value_strategy ENUM('same_as_start', 'last_meter_value', 'start_of_next_tx'),
  termination_reason TEXT NULL
);
```

`transaction_pk` references the id in `transaction` table. `terminated_at` and `termination_reason` are self-explanatory.

`stop_timestamp_strategy` describes how the value is selected in the following fallback order:

- `last_meter_value`: if a subsequent transaction at this chargebox/connector exists, find the last sent meter value for this transaction before the subsquent transaction's `start_timestamp` 
- `start_of_next_tx`:  ok, there is no meter value. use the value of subsequent transaction's `start_timestamp` (if it exists)
- `same_as_start`: ok, there is no subsequent transaction. set the end value to the value of start.

`stop_value_strategy` describes how the value is selected in the following fallback order:
- `last_meter_value`: _same as above_
- `start_of_next_tx`: some charging stations do not reset the meter value counter (our case) after each transaction and continue counting. in such cases, use the value of subsequent transaction's `start_value` (if it exists)
- `same_as_start`: _same as above_

now, after the transaction is manually terminated, a charging station might send StopTransaction for it. in this case, charging station's action has priority and we rollback the termination. i'm not sure yet whether we should just delete the entry in `transaction_terminated` or keep it and store this action as well (for example another field such as `termination_rollback_at`).

**Comment by chuck-h** [comment]
Slight variation on @goekay thought. Suppose the new db table is used for _every_ stop event (both "normal" and "manually terminated" events). If multiple stop events are associated with a single transaction, they can all be accessed from the transaction detail screen.

**Comment by goekay** [comment]
@chuck-h thanks, this is a good idea. i will take this further and split `transaction` table into two: `transaction_start` and `transaction_stop`. by doing this, the tables will just write timestamped events as in `connector_meter_value` and `connector_status`. this will make reading logic a little bir more complicated, though.

**Comment by chuck-h** [comment]
Does this also bear on handling of transactions initiated by local authorization while the CP is offline? I seem to recall seeing that some CP's generate a default transaction id (e.g. 0) under this circumstance.

**Comment by goekay** [comment]
i have committed an initial impl with 61ffb89. please review the code (most importantly [how the value/timestamp selection happens](https://github.com/RWTH-i5-IDSG/steve/blob/61ffb8944cc690c0e3b48e99765aea47095086f5/src/main/java/de/rwth/idsg/steve/service/TransactionTerminationService.java#L57)) and test it in a _test environment_. this impl has the following conscious decisions:
- termination of multiple transactions not possible (_yet_). i need more frontend know-how to develop this. someone with experience is welcome to work on this.
- manual termination can happen only once. however, charging station can overwrite the values, if it sends a StopTransaction later. the data that the station sends will be preferred as stop value/timestamp.

**Comment by goekay** [comment]
> Does this also bear on handling of transactions initiated by local authorization while the CP is offline? I seem to recall seeing that some CP's generate a default transaction id (e.g. 0) under this circumstance.

the behaviour did not change in this regard. so, no. the transaction id is still tied to the primary key in `transaction_start` (former `transaction`) and there can only be at most one of 0.

**Comment by goekay** [comment]
since i merged the [feature branch](https://github.com/RWTH-i5-IDSG/steve/tree/terminate-transactions) into master, i am closing this issue.
